### DISCUSSION TITLE
Reconsider using LibGit2Sharp?

### DISCUSSION THREAD

**Discussion by cidzoo** [adamralph/minver#974] [Show and tell]
As I now understand, initially the project was using _Lib2GitSharp_ (#256 in 2019) and as been moved to use the (if available) local git installation. 

Funnily, without knowing this, I've forked the project a few days ago and try to use Lib2Git (again). 

## Rationals

My initial needs is to not rely on a local external git dependency. Indeed, I'm doing CI and using some docker images to build my projects. However, they may not all have a git installed. Therefore I has to create custom images for thoses to integrate git. Aside the fact that it was not easy for some of them, it also implies I need to maintain them, rebuild them for new version, etc. 

So  the main rational for me is to remove an external dependency and thus make _MinVer_ self-contained.

 ## What I've done

So, as stated above, I've already forked the project and tested the switch. It's working fine and all tests are passing except one, but I suspect it is because this particular test is expecting a local git.

## My point of view

From my point of view, I think it really make sense to reuse existing code and not reinventing the wheel. To me, the current implementation looks quite "hacky" honestly. Parsing directly some command output is always a bit risky, some substring index are hardcoded, etc. When I switched back to _LibGit2Sharp_, implementation has been greatly simplified, it was as simple as using a few methods and properties of the lib without any hacky stuff. 

## Open questions

So I'd like to know if it's still relevant to keep using a local git installation instead of _Lib2GitSharp_ now in 2024? Are the aspects mentioned in the original PR still valid? And also what's your feeling about it?

- #149 with _Lib2GitSharp_ `v0.26.0`
  - Now version is `v0.29.0` but the related issue https://github.com/libgit2/libgit2sharp/issues/1647 is still open, so not sure if it has been fixed or not
- #269 
  - issue description is very broad, so no sure to understand why it's relevant and also what was the issue with _LibGit2Sharp_ in this context
- Now that we are 5 years later, do you guys think is it worth reconsidering the use of _LibGit2Sharp_?
  - if yes, I'm willing to continue my investigations and provide a PR, also maybe testing the cases mentioned about Alpine, etc.

**Comment by adamralph** [comment]
Also see https://github.com/adamralph/minver/issues/940

**Reply by cidzoo** [reply]
Ok thanks. I've checked #940 and linked/related issues. Though, not sure to understand what's your plan and position about this today?

My understanding is that you're now revaluting the use of MSBuild task after having decided not to use it in 2019 as described in #112. Now #112 has been reopen after the consideration presented in #940. By the way, why keeping both #112 and #940 opened? It may be misleading for a newcomer.

Currently a draft PR #971 addresses this. It's also using back _LibGit2Sharp_ apparently. That's what you meant? What's the current status on this PR? Is it planned for v5? And also when is v5 release date? 

If this is indeed addressing my concern about git, meaning that v5 will be self-contained again, I will close this discussion as resolved.

Thanks for your help and sorry if I've missed something. Oh and great piece of lib, it has been very useful ;)

**Reply by adamralph** [reply]
> what's your plan and position about this today?

No plan or position currently.

> By the way, why keeping both #112 and #940 opened? It may be misleading for a newcomer.

I've converted #940 to a discussion.

> Currently a draft PR https://github.com/adamralph/minver/pull/971 addresses this. It's also using back _LibGit2Sharp_ apparently.

#971 does not use LibGit2Sharp. It's a draft PR to continue the discussion about what's required to switch to an MSBuild task.

> What's the current status on this PR?

It's a draft PR to continue the discussion about what's required to switch to an MSBuild task.

> Is it planned for v5?

No

> And also when is v5 release date?

No planned release date.

**Reply by cidzoo** [reply]
> https://github.com/adamralph/minver/pull/971 does not use LibGit2Sharp. It's a draft PR to continue the discussion about what's required to switch to an MSBuild task.

Yes indeed my bad, I didn't realized it wasn't in the PR anymore. But I had the confusion because you've cherry-picked some of the commits from @bording branch (https://github.com/adamralph/minver/pull/971#issuecomment-1896290864) but left behind [the commit](https://github.com/bording/minver/commit/43e7b4fda9d8a14b7660fb93656b50bbd58ddc44) implementing `LibGit2Sharp`. Obviously this was on purpose, but I can't find the rationals in the PR discussion? 

From my understanding of https://github.com/adamralph/minver/discussions/976, the goal of switching back to MSBuild task is to improve performances? And @bording also mentions that using `LibGit2Sharp` further improve performances.

If switching back to `LibGit2Sharp` improves performances and remove an external system dependency (my initial consideration above), I think it seems worth reconsidering? If you still consider it's not, could you please argue why?

**Reply by adamralph** [reply]
While using `LibGit2Sharp` may give some performance improvements (although I suspect they would be minor, or even insignificant in most cases), it feels like the added cost of maintenance (e.g. testing on various Linux distributions and dealing with any associated bugs) outweighs them, and I'm closing this discussion on that basis.
